### PR TITLE
Add downtime module

### PR DESCRIPTION
**Pull request by zippolyte**
##### SUMMARY

Managing downtimes with the current datadog_monitor module is not really practical and the recommended solution anymore.

This adds the downtime module for managing monitor downtimes

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

`datadog_downtime`

### CODE REVIEW

File: plugins/modules/monitoring/datadog/datadog_downtime.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+DOCUMENTATION = """
+---
+module: datadog_downtime
+short_description: Manages Datadog downtimes
```
Comment (felixfontein) [inline plugins/modules/monitoring/datadog/datadog_downtime.py:14]:
```suggestion
short_description: Manages Datadog downtimes
version_added: 2.0.0
```

File: plugins/modules/monitoring/datadog/datadog_downtime.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+description:
+  - Manages downtimes within Datadog.
+  - Options as described on https://docs.datadoghq.com/api/.
```
Comment (felixfontein) [inline plugins/modules/monitoring/datadog/datadog_downtime.py:17]:
```suggestion
  - Options as described on U(https://docs.datadoghq.com/api/).
```

File: plugins/modules/monitoring/datadog/datadog_downtime.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+        elements: str
+    monitor_id:
+        description:
+          - The ID of the monitor to mute. If not provided, the downtime applies to all monitors.
+        type: int
+    downtime_message:
+        description:
+          - A message to include with notifications for this downtime.
+          - Email notifications can be sent to specific users by using the same "@username" notation as events.
+          - Monitor message template variables can be accessed by using double square brackets, i.e "[[" and "]]".
```
Comment (felixfontein) [inline plugins/modules/monitoring/datadog/datadog_downtime.py:72]:
```suggestion
          - Monitor message template variables can be accessed by using double square brackets, i.e C([[) and C(]]).
```

File: plugins/modules/monitoring/datadog/datadog_downtime.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+          - POSIX timestamp to end the downtime. If not provided, the downtime is in effect until you cancel it.
+    timezone:
+        description:
+          - The timezone for the downtime.
+        type: str
+    rrule:
+        description:
+          - The C(RRULE) standard for defining recurring events.
+          - For example, to have a recurring event on the first day of each month,
+          - select a type of rrule and set the C(FREQ) to C(MONTHLY) and C(BYMONTHDAY) to C(1).
```
Comment (felixfontein) [inline plugins/modules/monitoring/datadog/datadog_downtime.py:90]:
```suggestion
            select a type of rrule and set the C(FREQ) to C(MONTHLY) and C(BYMONTHDAY) to C(1).
```
You split up that sentence into two paragraphs.

File: plugins/modules/monitoring/datadog/datadog_downtime.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+      state: present
+      monitor_tags: ["foo:bar"]
+      downtime_message: "Downtime for foo:bar"
+      scope: "test"
+      api_key: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+      app_key: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+      # Lookup the id in the file and ignore errors if the file doesn't exits, so downtime gets created
+      id: "{{ lookup('file', inventory_hostname +'_downtime_id.txt', errors='ignore') }}"
+  - name: Save downtime id to file for later updates and idempotence
+    local_action: copy content="{{ downtime_var.downtime.id }}" dest={{ inventory_hostname +'_downtime_id.txt' }}
```
Comment (felixfontein) [inline plugins/modules/monitoring/datadog/datadog_downtime.py:109]: Use `delegate_to` instead of `local_action`. That also improves readability a lot.

**Review by felixfontein — COMMENTED** [review]
Please note that all new modules and plugins must have tests (unit tests, integration tests, or both). (Integration tests can also marked as `unsupported` if they cannot run in CI.)

**Comment by felixfontein** [comment]
Instead of adding more datadog modules here, you could also migrate them to a separate collection. That would allow you to set up a CI with working integration tests, and be a lot more flexible with releasing.

CC @gundalow

**Comment by DavidWittman** [comment]
> @!UNKNOWN @Akasurde @Alb0t @AnderEnder @ApsOps @AugustusKling @Aversiste @ColOfAbRiX @DavidWittman @Deepakkothandan @DenBeke @FlossWare @GR360RY @GwenaelPellenArkeup @InTheCloudDan @JayKayy @Jmainguy @JoergFiedler @KKoukiou @KellerFuchs @KimNorgaard @L2G @LinusU @Lunik @MacLemon @MatrixCrawler @MeganLiu @Mogztter @MorrisA @NickatEpic @NilashishC @nitaco @Nosmoht @Qalthos @QijunPan @QuentinBrosse @RickS-C137 @Rylon @Sajna-Shetty @SamyCoenen @Sedward @Shaps @Slezhuk @Smithx10 @Spredzy @Tatsh @ThePixelDeveloper @TimothyVandenbrande @TommyLike @Tomorrow9 @UnderGreen @Wolfant @aajdinov @aalexmonteiro @abarbare @abellotti @abulimov @adamgoossens @adamvaughan @adejoux @adriane-cardozo @adrianmoisey @agaffney @agmezr @ahtik @aimonb @akostyuk @albertomurillo @alcamie101 @alikins @alxgu @amasolov @aminvakil @amit0701 @andreparames @andrew-d @andsens @andyhky @andytom @angristan @angstwad @arturaz @azaghal @bachradsusi @baldwinSPC @bannaych @barnabycourt @barryib @bbyhuy @bcoca @bendoh @bennojoy @berenddeboer @berendt @bgaifullin @bgurney-rh @bhcopeland @bigmstone @billdodd @bincyber @bjolivot @bleader @bpennypacker @brampling @bregman-arie @brian-brazil @briceburg @brontitall @bsanders @bushvin @bvitnik @bwhaley @caphrim007 @carchi8py @catcombo @cben @cheese @chris93111 @chrishoffman @chrisisbeef @claco @clc-runner @clementtrebuchet @cloudnull @cmprescott @colin-nolan @commel @coreywan @cprh @cwollinger @d-little @danieljaouen @danielmellado @dankeder @dareko @davx8342 @dcermak @dch @decentral1se @derez @dermute @devyanikota @dinoocch @dirtyharrycallahan @displague @dj-wasabi @djmattyg007 @dkorn @dmtrs @dnix101 @dohoangkhiem @dougluce @dprts @drcapulet @drew-russell @drewkerrigan @drybjed @dstoflet @ebirn @edevenport @edisonxiang @eest @eikef @elad661 @elasticdog @enriclluelles @erjohnso @erydo @eryx12o45 @evertmulder @evgkrsk @evrardjp @fabulops @farhan7500 @fcuny @fgbulsoni @fishman @flaper87 @flynn1973 @fraff @freesky-edward @fxfitz @gamethis @ganeshrn @garbled1 @gautamphegde @genegr @getjack @gforster @giovannisciortino @giuseppe @glitchcrab @groks @gtanzillo @guillaume_ro_fr @haad @hekonsek @helldorado @hkariti @hnakamur @hnanni @hogarthj @hryamzik @huaweicloud @hulquest @hwDCN @ilicmilan @indrajitr @inetfuture @ivanvanderbyl @jagadeeshnv @jails @jairojunior @jake2184 @jamescassell @jasperla @jbenden @jcftang @jcgruenhage @jdauphant @jerome-quere @jhoekx @jirutka @jlaska @jle64 @johanwiren @john-westcott-iv @jose-delarosa @joshainglis @jparrill @jpdasma @jpmens @jsmartin @jsumners @jtyr @justjais @kahowell @kairoaraujo @kamsz @karmab @kavu @kbrebanov @keachi @kenichi-ogawa-1988 @kevensen @kindermoumoute @krisvasudevan @krsacme @kubevirt @kunkku @kyleabenson @lekum @levonet @lionmax @lmprice @lonico @lrupp @lukasbestle @m-yosefpor @machacekondra @madhav-bharadwaj @magnus919 @makaimc @manojmeda @marc-sensenich @markuman @marns93 @martinm82 @marvin-sinister @marwatk @matbu @matburt @mator @mattjeffery @mattupstate @matze @mavit @maxamillion @mcltn @mcodd @meerkampdvv @mekanix @melodous @mgedmin @mgruener @mheap @mmazur @molekuul @mpdehaan @mraineri @mross22 @mscherer @mulby @mwarkentin @n0trax @n0ts @nalsaber @nasx @nate-kingsley @natefoo @navalkp @nbuchwitz @ndavison @ndclt @ndswartz @nerzhul @neuroid @nibalizer @nitzmahone @niuzhenguo @noles @noseka1 @nurfet-becirevic @obirvalger @oboukili @obourdon @ogenstad @oolongbrothers @opoplawski @opslounge @orgito @ovcharenko @overhacked @parfeon @pascalheraud @pb8226 @pertoft @phumpal @pilou- @pkliczewski @pmakowski @pmarkham @prabhosa @precurse @privateip @pubnub @pyykkis @quentinsf @quidame @raekins @rajeevarakkal @rambleraptor @ramondelafuente @ramooncamacho @ravibhure @remixtj @remyleone @ribbons @ricardogpsf @rmcintosh @robinro @robwagner33 @rohitChaware @rosmo @rosowiecki @rsmontero @russoz @rvalle @ryansb @sac @samdoran @saranyasridharan @sayap @scathatheworm @schmots1 @scodeman @scottanderson42 @sdodsley @seandst @sermilrod @sganesh-infoblox @sgargan @shane-walker @shirou @sieben @sile16 @sivel @sjaiswal @skinp @skornehl @slok @sm4rk0 @smashwilson @smbambling @snopoke @soodpr @srvg @steamx @stearz @stpierre @stygstra @stympy @supertom @suprememoocow @sysadmind @szinck @t0mk @t794104 @tacatac @talzur @tarka @tastychutney @tbielawa @tchernomax @tdtrask @teebes @tgoetheyn @thaumos @the-maldridge @thoiberg @tintoy @tksmd @tmiotto @tmshn @toabctl @tomasg2012 @tonyseek @treyperry @trishnaguha @troy2914 @tumbl3w33d @turb @tuxillo @tyouxa @tzure @ujwalkomarla @usawa @vaygr @vcarceler @vedit @verkaufer @vexata @vfauth @vincentvdk @vmalloc @vritant @waheedi @wbrefvem @weaselkeeper @willthames @willybarro @wltjr @wopfel @wtcross @xcambar @xen0l @xiaozhu36 @xorel @xuxiaowei0512 @yaacov @yanzhangi @yeukhon @zanssa @zbal @zengchen1024 @zgalor @zhhuta @zhongjun2 @zimbatm
> 
> As a maintainer of a module in the same namespace this new module has been submitted to, your vote counts for shipits. Please review this module and add `shipit` if you would like to see it merged.
> 
> [click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

This is insane. How do I remove myself from these after the collections change?

**Comment by felixfontein** [comment]
@DavidWittman in theory you shouldn't be pinged because you are on the ignore list (https://github.com/ansible-collections/community.general/blob/main/.github/BOTMETA.yml#L804) for the only module where you are listed as author (https://github.com/ansible-collections/community.general/blob/main/plugins/modules/system/alternatives.py#L20). It looks to me that there is a bug in the bot. Also, the module you're listed as an author is in plugins/modules/systems/, not in plugins/modules/monitoring. Could be another bug (caused by the symlinks).

@gundalow can you please take a look?

**Comment by DavidWittman** [comment]
> @DavidWittman in theory you shouldn't be pinged because you are on the ignore list (https://github.com/ansible-collections/community.general/blob/main/.github/BOTMETA.yml#L804) for the only module where you are listed as author (https://github.com/ansible-collections/community.general/blob/main/plugins/modules/system/alternatives.py#L20). It looks to me that there is a bug in the bot. Also, the module you're listed as an author is in plugins/modules/systems/, not in plugins/modules/monitoring. Could be another bug (caused by the symlinks).
> 
> @gundalow can you please take a look?

@felixfontein Thank you for the thoughtful response to my moment of temporary rage. 👍

**Comment by zippolyte** [comment]
Hmm question: the datadog library used in the module is py3 only. Is that a problem ? If not, how do I specify to run tests for py3 only ?

File: tests/unit/requirements.txt [hunk 6]
```diff
@@ -21,3 +21,6 @@ openshift ; python_version >= '2.7'
 # requirement for maven_artifact module
 lxml
 semantic_version
+
+# requirement for datadog_downtime module
+datadog-api-client >= 1.0.0b3
```
Comment (felixfontein) [inline tests/unit/requirements.txt:26]:
```suggestion
datadog-api-client >= 1.0.0b3 ; python_version >= '3.5'
```

File: tests/unit/plugins/modules/monitoring/test_datadog_downtime.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+from ansible_collections.community.general.plugins.modules.monitoring.datadog import datadog_downtime
+from ansible_collections.community.general.tests.unit.compat.mock import MagicMock, patch
+from ansible_collections.community.general.tests.unit.plugins.modules.utils import (
+    AnsibleExitJson, AnsibleFailJson, ModuleTestCase, set_module_args
+)
+from datadog_api_client.v1.model.downtime import Downtime
```
Comment (felixfontein) [inline tests/unit/plugins/modules/monitoring/test_datadog_downtime.py:13]: If you use `pytest.importorskip` here, this shouldn't fail when the library is not installed.

File: plugins/modules/monitoring/datadog/datadog_downtime.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+      scope: "test"
+      api_key: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+      app_key: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+      # Lookup the id in the file and ignore errors if the file doesn't exits, so downtime gets created
+      id: "{{ lookup('file', inventory_hostname +'_downtime_id.txt', errors='ignore') }}"
+  - name: Save downtime id to file for later updates and idempotence
+    delegate_to: localhost
+    copy:
+      content: "{{ downtime.downtime.id }}"
+      dest: "{{ inventory_hostname +'_downtime_id.txt' }}"
```
Comment (felixfontein) [inline plugins/modules/monitoring/datadog/datadog_downtime.py:112]: Jinja2's preferred string concatenation operator is `~`:
```suggestion
      dest: "{{ inventory_hostname ~ '_downtime_id.txt' }}"
```

File: plugins/modules/monitoring/datadog/datadog_downtime.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+    register: downtime_var
+    community.general.datadog_downtime:
+      state: present
+      monitor_tags: ["foo:bar"]
+      downtime_message: "Downtime for foo:bar"
+      scope: "test"
+      api_key: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+      app_key: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+      # Lookup the id in the file and ignore errors if the file doesn't exits, so downtime gets created
+      id: "{{ lookup('file', inventory_hostname +'_downtime_id.txt', errors='ignore') }}"
```
Comment (felixfontein) [inline plugins/modules/monitoring/datadog/datadog_downtime.py:107]:
```suggestion
      id: "{{ lookup('file', inventory_hostname ~ '_downtime_id.txt', errors='ignore') }}"
```

File: plugins/modules/monitoring/datadog/datadog_downtime.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+          - Attributes specifying the duration in C(RRULE) are not supported (e.g. C(DTSTART), C(DTEND), C(DURATION)).
+        type: str
+"""
+
+EXAMPLES = """
+  - name: Create a downtime
+    register: downtime_var
+    community.general.datadog_downtime:
+      state: present
+      monitor_tags: ["foo:bar"]
```
Comment (felixfontein) [inline plugins/modules/monitoring/datadog/datadog_downtime.py:101]: The 'long' YAML syntax might be easier to read:
```suggestion
      monitor_tags:
        - "foo:bar"
```

File: plugins/modules/monitoring/datadog/datadog_downtime.py [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+---
+module: datadog_downtime
+short_description: Manages Datadog downtimes
+version_added: 2.0.0
+description:
+  - Manages downtimes within Datadog.
+  - Options as described on U(https://docs.datadoghq.com/api/v1/downtimes/s).
+author:
+  - Datadog (@Datadog)
+requirements: [datadog-api-client]
```
Comment (felixfontein) [inline plugins/modules/monitoring/datadog/datadog_downtime.py:21]: Make the Python 3 dependency a bit more explicit:
```suggestion
requirements:
  - datadog-api-client
  - Python 3
```

File: plugins/modules/monitoring/datadog/datadog_downtime.py [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+        type: int
+        description:
+          - POSIX timestamp to start the downtime. If not provided, the downtime starts the moment it is created.
+    end:
+        type: int
+        description:
+          - POSIX timestamp to end the downtime. If not provided, the downtime is in effect until you cancel it.
+    timezone:
+        description:
+          - The timezone for the downtime.
```
Comment (felixfontein) [inline plugins/modules/monitoring/datadog/datadog_downtime.py:84]: Out of curiosity: for what is the timezone needed? To decide which time to show to the user?

Comment (zippolyte) [reply plugins/modules/monitoring/datadog/datadog_downtime.py:84]: Yes it's only used in the UI for showing to the user.

**Review by felixfontein — APPROVED** [review]
Looks good to me.

**Comment by felixfontein** [comment]
@zippolyte thanks for this contribution!

**Pull request merged**